src/src/PluginStructs/P014_data_struct.cpp

```diff
@@ -175,148 +249,279 @@ uint8_t P014_data_struct::checkCRC(uint16_t data, uint8_t check)
   return (uint8_t)remainder;
 }
 
+
+
+
+
 /* ======================================================================
-   Function: si7021_readRegister
-   Purpose : read the user register from the sensor
-   Input   : user register value filled by function
-   Output  : 0 if okay
-   Comments: -
-   ====================================================================== */
-int8_t P014_data_struct::readRegister(uint8_t *value)
+Function: Plugin_123_si7013_setResolution
+Purpose : Sets the sensor resolution to one of four levels
+Input   : see #define default is SI7013_RESOLUTION_14T_12RH
+Output  : 0 if okay
+Comments: -
+====================================================================== */
+bool P014_data_struct::setResolution(uint8_t i2caddr, uint8_t resolution)
 {
-  // Request user register
-  Wire.beginTransmission(SI7021_I2C_ADDRESS);
-  Wire.write(SI7021_READ_REG);
-  Wire.endTransmission();
-
-  // request 1 uint8_t result
-  Wire.requestFrom(SI7021_I2C_ADDRESS, 1);
+  bool ok;
+  uint8_t reg = I2C_read8_reg(i2caddr,SI70xx_CMD_READ_REG1,&ok);
+  if (ok) {
+    // remove resolution bits
+    reg &= SI70xx_RESOLUTION_MASK ;
+    // Write the new resolution bits but clear unused before
+    ok = I2C_write8_reg(i2caddr,SI70xx_CMD_WRITE_REG1, reg | ( resolution &= ~SI70xx_RESOLUTION_MASK) );
+  }
 
-  if (Wire.available() >= 1) {
-    *value = Wire.read();
-    return 0;
+  if (!ok) {
+      String log = F("SI70xx : Unable to set resolution: ");
+      log += String(i2caddr, HEX);
+      addLog(LOG_LEVEL_ERROR, log);
   }
 
-  return 1;
+  return ok;
 }
 
-/* ======================================================================
-   Function: startConv
-   Purpose : return temperature or humidity measured
-   Input   : data type SI7021_READ_HUM or SI7021_READ_TEMP
-   Output  : 0 if okay
-   Comments: -
-   ====================================================================== */
-uint8_t P014_data_struct::startConv(uint8_t datatype)
-{
-  // Request a reading
-  Wire.beginTransmission(SI7021_I2C_ADDRESS);
-  Wire.write(datatype);
+
+inline bool P014_data_struct::softReset(uint8_t i2caddr){
+  // Prepare to write to the register value
+    bool ret = I2C_write8(i2caddr, SI70xx_CMD_SOFT_RESET);
+    delay(50); //MFD: check the datasheet
```

Comment:
Why do we need a delay here?
Why not add a state to the state machine to eliminate this delay?

Reply:
I still have to check the datasheet on this across all devices to see if it is even required.
I don't think is required at all.
If required I will implement an extra state for it. otherwise I will take it out after testing.

Reply:
done.